sysdeps/generic-helpers/include/generic-helpers/netlink.hpp

```diff
@@ -0,0 +1,294 @@
+#pragma once
+
+#include <ifaddrs.h>
+#include <linux/rtnetlink.h>
+#include <net/if.h>
+#include <netpacket/packet.h>
+#include <sys/socket.h>
+
+#include <frg/allocation.hpp>
+#include <frg/span.hpp>
+
+#include "mlibc/allocator.hpp"
+#include "mlibc/posix-sysdeps.hpp"
+
+namespace {
+
+/**
+ * Helper class for issuing rtnetlink requests and receiving replies.
+ */
+struct NetlinkHelper {
+	NetlinkHelper() {
+		data_ = reinterpret_cast<uint8_t *>(getAllocator().allocate(size_));
+	}
+
+	~NetlinkHelper() {
+		getAllocator().free(data_);
+	}
+
+	bool send_request(int type) {
+		if(!fd_ || *fd_ == -1) {
+			int fd;
+			if(int e = mlibc::sys_socket(AF_NETLINK, SOCK_RAW | SOCK_CLOEXEC, NETLINK_ROUTE, &fd); e)
+				return false;
+			fd_ = fd;
+		}
+
+		struct NetlinkMessage {
+			nlmsghdr hdr;
+			rtgenmsg msg;
+		} request;
+
+		memset(&request, 0, sizeof(request));
+		request.hdr.nlmsg_flags = NLM_F_DUMP | NLM_F_REQUEST;
+		request.hdr.nlmsg_type = type;
+		request.hdr.nlmsg_len = NLMSG_LENGTH(sizeof(request) - sizeof(nlmsghdr));
+		request.msg.rtgen_family = AF_UNSPEC;
+
+		return send(fd_.value(), reinterpret_cast<void *>(&request), sizeof(request), 0) == sizeof(request);
+	}
+
+	bool recv(void cb(void *, nlmsghdr *), void *ctx) {
+		ssize_t read;
+
+		while((read = ::recv(fd_.value(), reinterpret_cast<msghdr *>(data_), size_, 0)) > 0) {
+			auto hdr = reinterpret_cast<nlmsghdr *>(data_);
+
+			for(; NLMSG_OK(hdr, static_cast<size_t>(read)); hdr = NLMSG_NEXT(hdr, read)) {
+				if(hdr->nlmsg_type == NLMSG_DONE)
+					return true;
+				if(hdr->nlmsg_type == NLMSG_ERROR) {
+					nlmsgerr *err = reinterpret_cast<nlmsgerr *>(NLMSG_DATA(hdr));
+					mlibc::infoLogger() << "mlibc: NetlinkHelper received NLMSG_ERROR " << -err->error << ": "
+						<< strerror(-err->error) << frg::endlog;
+					return false;
+				}
+
+				cb(ctx, hdr);
+			}
+		}
+
+		return false;
+	}
+private:
+	frg::optional<int> fd_;
+	size_t size_ = 8192;
```

Comment:
If this is always set to 8192 then we don't need a heap allocation right? And otherwise I'd much rather see a ``frg::vector`` be used here.